Add iSCSI test suite

KyleMahlkuch opened the pull request:
I've made some tests that test iSCSI specific configs. These tests are based off of existing network and disk tests so I've titled my tests [old_test]_iscsi.py. Hopefully this makes it easier to compare to old tests and understand their functionality. Please review and let me know if more work needs to be done on these tests. 
This is just a starting point to enable basic testing for iSCSI. Hopefully more tests will be added in the future.
I've also made a config file for the fvt-avocado-wrapper which I will submit once my iscsi tests are added to this repo.

Thanks,
Kyle

narasimhan-v commented:
@KyleMahlkuch Thanks for the patch.

I have a generic pointer.

1. If you want to have these tests doing the same thing as existing ones, then we are going for code duplication. Would it not be better to have existing ones working for iscsi usecase ? We can work together and get it done as well. I see that the change you have in major is yaml, which can be dealt with easily as we have done that for multiple cases, including for fiotest.

KyleMahlkuch commented:
Hi Narasimhan, 
Yes, it would be easier to combine these changes into the current tests. 
In addition to yaml changes I changed some of the file system creation code as it was not working correctly on my setup. I'm not sure if this is an iscsi issue or an error on my part. We could disable this code for iscsi in the mean time.
I also had to make some changes to the disk identification code. The current disk tests were unable to identify the iscsi disks for some reason. Like you suggested over slack we could set a flag for iSCSI through yaml that would do some additional parsing of disks.

narasimhan-v commented:
Sure @KyleMahlkuch please send in your changes, and we can see if that can be applied in generic case too.

KyleMahlkuch commented:
Hi @narasimhan-v, looks like I was running with some old code. I checked out the latest version and this makes things much simpler for iscsi tests. I added a ethtool yaml file and corrected some fio issues I ran into. I believe the fio issues may be universal but I can make these changes specific to iscsi runs if you think that's the correct approach.  Thanks for your help!

narasimhan-v commented on io/disk/fiotest.py:
```diff
@@ -212,7 +214,7 @@ def test(self):
         self.log.info("Test will run on %s", self.dirs)
         fio_job = self.params.get('fio_job', default='fio-simple.job')
-        cmd = '%s/fio %s %s --filename=%s' % (self.sourcedir,
+        cmd = '%s/fio %s --filename=%s%s' % (self.sourcedir,
```
@KyleMahlkuch Can you please explain why this particular change is required ?

KyleMahlkuch replied:
Example from my run on iscsi: 
`/var/tmp/avocado_4rvzt9du/fio/fio /root/avocado-fvt-wrapper/tests/avocado-misc-tests/io/disk/fiotest.py.data/fio-simple.job /dev/mapper/mpatha --filename=fiotest-image`
The "no_fs" tests failed with the above command. filename should point to the mpath so that io is going across the iscsi connection. 

My process for running fio without a filesystem is:
/path/to/fio /path/to/job --filename=/path/to/disk
According to the fio man page: 
fio [options] [jobfile]

So my commit is wrong. It should be: `'%s/fio %s --filename=%s'` because I want filename to be `/dev/mapper/mpatha`
Maybe for the "no_fs" runs there should be a seperate fio command line.

narasimhan-v replied:
Ok, getting your idea. The no_fs option has been working with existing code, but not consistently enough.
I will give this one a try, probably Monday, and then merge this one.

If you have a log on which this works fine, please post one so that it also serves as a comparison point.

narasimhan-v replied:
Feel free to update the commit here, and also take care of indentation in the next 2 lines.

narasimhan-v commented:
@KyleMahlkuch 
Indentation issue and commit message format needs to be taken care of, as per https://travis-ci.org/github/avocado-framework-tests/avocado-misc-tests/jobs/711525704

KyleMahlkuch commented:
> @KyleMahlkuch
> Indentation issue and commit message format needs to be taken care of, as per https://travis-ci.org/github/avocado-framework-tests/avocado-misc-tests/jobs/711525704

Sorry about all the pushes but this is now fixed.

KyleMahlkuch commented:
@narasimhan-v 
I made separate logic for fs and no_fs runs since no_fs does not require the filename, just a path to disk. 
With my patch fio throws a status of 3 when running on no_fs iscsi disks. Fio process completes and gives good output
[fio_iscsi_patched_job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/4983173/fio_iscsi_patched_job.log)
 so I figured this would be better as a warning than a fail. If this change is too specific we can remove it. I'll have to investigate what is going wrong here. 
I am attaching a job log with the latest version of this commit (plus some additional logging)

[fio_iscsi_patched_job.log](https://github.com/avocado-framework-tests/avocado-misc-tests/files/4983175/fio_iscsi_patched_job.log)
.

KyleMahlkuch commented:
I corrected these formatting errors but travis is still failing. Looks like the pull request is behind my fork?

KyleMahlkuch commented:
Hi @narasimhan-v have you had a chance to test these changes? Thanks.

narasimhan-v commented:
@KyleMahlkuch Apologies, this slipped of my radar.

I tested this with an NVMe disk and I see all `no_fs` tests are warning. Previously, only the `no_fs no_raid no_lv` case was failing.
I am still not sure if this is the correct way to do.
```
# avocado run fiotest.py -m fiotest.py.data/fio.yaml
JOB ID     : b2e7aeac2bbb0d213eb6096f5c1c861c17929224
JOB LOG    : /root/sim/tests/results/job-2020-08-04T06.05-b2e7aea/job.log
 (1/8) fiotest.py:FioTest.test;run-fs-xfs-lv-lv-raid-raid-9d29: PASS (54.83 s)
 (2/8) fiotest.py:FioTest.test;run-fs-xfs-lv-lv-raid-no_raid-d398: PASS (38.61 s)
 (3/8) fiotest.py:FioTest.test;run-fs-xfs-lv-no_lv-raid-raid-ebb2: PASS (36.74 s)
 (4/8) fiotest.py:FioTest.test;run-fs-xfs-lv-no_lv-raid-no_raid-bfea: PASS (36.54 s)
 (5/8) fiotest.py:FioTest.test;run-fs-no_fs-lv-lv-raid-raid-472c: WARN: Test passed but there were warnings during execution. Check the log for details. (34.96 s)
 (6/8) fiotest.py:FioTest.test;run-fs-no_fs-lv-lv-raid-no_raid-b988: WARN: Test passed but there were warnings during execution. Check the log for details. (34.65 s)
 (7/8) fiotest.py:FioTest.test;run-fs-no_fs-lv-no_lv-raid-raid-13d4: WARN: Test passed but there were warnings during execution. Check the log for details. (31.38 s)
 (8/8) fiotest.py:FioTest.test;run-fs-no_fs-lv-no_lv-raid-no_raid-756e: WARN: Test passed but there were warnings during execution. Check the log for details. (33.42 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 4 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/sim/tests/results/job-2020-08-04T06.05-b2e7aea/results.html
JOB TIME   : 302.16 s
```

narasimhan-v commented on io/disk/fiotest.py:
```diff
@@ -212,11 +214,26 @@ def test(self):
-            self.fail("fio run failed")
+
+	# if fs is present create a file on that fs, if no fs 
+	# self.dirs = path to disk, thus a filename is not needed
```
s/filename/fio_file/

KyleMahlkuch commented:
Thanks for testing @narasimhan-v. Could you share the logs?

narasimhan-v commented:
> Thanks for testing @narasimhan-v. Could you share the logs?

[job_log.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/5022709/job_log.txt)

KyleMahlkuch commented:
`seq-readwrite: (groupid=6, jobs=1): err=84 (file:io_u.c:2069, func=io_u_sync_complete, error=Invalid or incomplete multibyte or wide character): pid=16614: Tue Aug  4 06:10:13 2020`
This is the same error being seen on your run and my iscsi run. 
I'm not sure the reason for the error but adding `time_based=1` to the fio job seems to fix this issue. 
There is an issue open for this in the fio github here: https://github.com/axboe/fio/issues/740

narasimhan-v commented:
> `seq-readwrite: (groupid=6, jobs=1): err=84 (file:io_u.c:2069, func=io_u_sync_complete, error=Invalid or incomplete multibyte or wide character): pid=16614: Tue Aug 4 06:10:13 2020`
> This is the same error being seen on your run and my iscsi run.
> I'm not sure the reason for the error but adding `time_based=1` to the fio job seems to fix this issue.
> There is an issue open for this in the fio github here: [axboe/fio#740](https://github.com/axboe/fio/issues/740)

@KyleMahlkuch Thanks. I guess we can merge this then. Can you include the fio issue link also in the commit message please ?

KyleMahlkuch commented:
> I'm not sure the reason for the error but adding `time_based=1` to the fio job seems to fix this issue.
I spoke too soon here. This does not correct the issue in all cases.

KyleMahlkuch commented:
Looks like the formatting issues are fixed. I added the fio issue to the commit.

KyleMahlkuch commented:
Thanks Narasimhan!
